_episodes/01-raster-structure.md

```diff
@@ -15,10 +15,474 @@ objectives:
+data at the edges of rasters.
+
+By default the shape of a raster is always rectangular. So if we have a dataset
+that has a shape that isn't rectangular, some pixels at the edge of the raster
+will have no data values. This often happens when the data were collected by an
+sensor which only flew over some part of a defined region.
+
+In the RGB image below, the pixels that are black have no data values. The sensor
+did not collect data in these areas.
+
+~~~
+rgb_HARV = rasterio.open("NEON-DS-Airborne-Remote-Sensing/HARV/RGB_Imagery/HARV_RGB_Ortho.tif", "r")
+rgb_HARV_arr = rgb_HARV.read()
+earthpy.plot.plot_rgb(
+    rgb_HARV_arr,
+    title="RGB Image, NoData Values UnMasked",
+    figsize=(10, 6)
+)
+~~~
+{: .language-python}
+
+
+<img src="../fig/01-demonstrate-no-data-black-3.png" title="plot of chunk demonstrate-no-data-black" alt="plot of chunk demonstrate-no-data-black" width="612" style="display: block; margin: auto;" />
+
+`rasterio` assigns a specific number as missing data to the `meta` attribute when the dataset is read, base don the file's own metadata. While the GeoTiff's `nodata` attribute is assigned to the value `-1.7e+308`, it turns out the missing data at the edges are represented by the value `0`. In order to run calculations on this image that ignore these edge values or plot he iamge without the nodata values being displayed on the color scale, we can mask out `0` values in our numpy array. 
+
+In the next image, the black edges have been masked using `numpy.ma.masked_where()`, a function that assigns no data values where a condition is true.
+
+~~~
+rgb_HARV_masked_arr = np.ma.masked_where(rgb_HARV_arr==0, rgb_HARV_arr) #1st argument is the condition, second is the array to mask
+earthpy.plot.plot_rgb(
+    rgb_HARV_masked_arr,
+    title="RGB Image, NoData Values Masked",
+    figsize=(10, 6)
+)
+~~~
+{: .language-python}
+
+<img src="../fig/01-demonstrate-no-data-masked-4.png" title="plot of chunk demonstrate-no-data-masked" alt="plot of chunk demonstrate-no-data-masked" width="612" style="display: block; margin: auto;" />
+
+The difference here shows up as ragged edges on the plot, rather than black
+spaces where there is no data.
+
+If your raster already has `nodata` values set correctly but you aren't sure where they are, you can deliberately plot them in a particular colour. This can be useful when checking a dataset's coverage. For instance, sometimes data can be missing where a sensor could not 'see' its target data, and you may wish to locate that missing data and fill it in. With Python, we can plot a boolean array of `True/False` values from our masked array's `.mask` attribute. Since the mask array represents no data values as `True` and data values as `False`, we need to reverse our boolean array so that we can more clearly see where no data values have been masked in each of our color channels.
+
+~~~
+earthpy.plot.plot_rgb(
```

Comment:
This code cries about two issues:

1. `name 'rgb_HARV_nan_arr' is not defined`
2. If you replace that with `rgb_HARV_masked_arr`, then `name 'ax' is not defined`

Reply:
Fixed, thanks for catching this. I tested the whole lesson again from scratch and it works.